protocols/ML01.md

```diff
@@ -0,0 +1,315 @@
+# ML01: Machine learning basic specification
+**THIS IS A DRAFT SPECFICATION AND IS NOT FINAL**
+
+| name                       | ML01               |
+|----------------------------|--------------------|
+| version                    | 0.1.0              |
+| status                     | proposal           |
+| author                     | Ryan Wilson (rw86) |
+| serving component(s)       | Machine Learning   |
+| consuming component(s)     | HCI                |
+| basic spec                 | yes                |
+| can be required by servers | no                 |
+| can be required by clients | HCI                |
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.
+
+
+## Changelog
+### Version 0.1
+
+- First draft
+
+# Foreword: Acknowledgements
+
+This specification is loosely based on ideas presented in an endpoint outline "ml_endpoints.md" written by Daphne Bogosian. There are some major architectural differences, outlined below:
+
+- The principal "backend should be responsible for all matters of persistance" agreed during protocol meetings has been incorporated
+  - In particular results are returned to the frontend as files via backend (this was also agreeed)
+- The principal "HCI should talk directly to backend" has been incorporated.
+  - In particular there is no BE <-> ML communication.
```

Comment:
Is this a mistake? I thought we agreed there would be no HCI <-> ML communications, only HCI -> BE -> ML?

Reply:
Checking back against the minutes it seems so. Thinking about how I would change it to be HCI -> BE -> ML I think (at least for this design) it makes more sense to me to "cut out the middle man" although, I didn't realise that this was deviating behaviour - mea culpa.

The HCI -> BE -> ML version would, basically involve BE providing an endpoint and forwarding requests and responses - so the protocol wouldn't change except all the URLs looking like "backend.endpoint/ML/...". I still do feel that BE kind of has a lot to do as it is - so to me adding any more intelligence to backend rather than "forwarding the requests" would feel like poor separation of concerns.

Reply:
I thought the idea of the HCI -> BE -> ML structure was that the calls made for ML (to BE from HCI) were pretty much just forwarded through to ML (after checking tokens), so shouldn't add to much work for BE. It does make BE a kind of useless middle man in that particular situation, but that was discussed and people still went with the idea, as it greatly simplifies the structure, so I think we should stick with it? (Not sure what you mean by adding more intelligence to BE, cause my understanding was they pretty much just were forwarding requests)?

Reply:
By intelligence I simply meant anything but forwarding. We seem to be on the same page on this. Except I am not too sure how ML and BE would authenticate the requests to each in this case? would ML still use the users token?

Reply:
Well I had this idea that BE would check that the user had permissions for the files HCI asked to ML. So if they didn't, the request would never even reach ML, just get an error code sent back. That way ML doesn't need to worry about tokens - if anything ever reaches them, they know there should be no issues with permissions?

Reply:
It also just occurred to me that BE would have to do a little bit more work when the user asks to train a model in that they will need to first check if that model exists already or not. But I think this would need to be done somehow no matter what?